<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Public Maps &#8212; Tumor Map Help</title>
    
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/custom.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Attribute Information" href="attrInfo.html" /> 
  </head>
  <body role="document">
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="index.html">Tumor Map Help</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="attrInfo.html" title="Attribute Information"
             accesskey="P">previous</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="public-maps">
<h1>Public Maps<a class="headerlink" href="#public-maps" title="Permalink to this headline">¶</a></h1>
<div class="section" id="pancan12">
<span id="pancan12-genemap"></span><span id="pancan12-samplemap"></span><h2>Pancan12<a class="headerlink" href="#pancan12" title="Permalink to this headline">¶</a></h2>
<p>The diverse tumor set called “Pan-Cancer-12,” is composed of 12 different
malignancies. It comprises 3,527 cases assayed by at least four of the six
possible data types routinely generated by The Cancer Genome Atlas: whole-exome
DNA sequence (Illumina HiSeq and GAII), DNA copy number variation (Affymetrix
6.0 microarrays), DNA methylation (Illumina 450,000-feature microarrays),
genome-wide mRNA levels (Illumina mRNA-seq), microRNA levels (Illumina
microRNA-seq), and protein levels for 131 proteins and/or phosphorylated
proteins (Reverse Phase Protein Arrays; RPPA).</p>
<p>The Sample Map layouts are composed of tumor samples collected in the study.
Attributes that can be explored on the Sample Map are described at
<a class="reference internal" href="#pancan12-samplemapattributes"><span class="std std-ref">Pancan12/SampleMap Attributes</span></a>.</p>
<p>The Gene Map layouts are composed of genes mapping to a probe from the data
collection platforms. E.g. MYC on the mRNA Gene Map, corresponds to the probe
mapping to MYC on the micro array platform. Attributes that can be explored on
the Gene Map are described at
<a class="reference internal" href="#pancan12-genemapattributes"><span class="std std-ref">Pancan12/GeneMap Attributes</span></a>.</p>
<p>Hoadley,K.A., Yau,C., Wolf,D.M., Cherniack,A.D., Tamborero,D., Ng,S.,
Leiserson,M.D.M., Niu,B., McLellan,M.D., Uzunangelov,V., et al. (2014)
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4152462/">Multiplatform analysis of 12 cancer types reveals molecular classification
within and across tissues of origin</a>.
Cell, 158, 929–44.</p>
</div>
<div class="section" id="gliomas">
<span id="id1"></span><h2>Gliomas<a class="headerlink" href="#gliomas" title="Permalink to this headline">¶</a></h2>
<p>We assembled a dataset comprising all TCGA newly diagnosed diffuse glioma
consisting of 1,122 patients and comprehensively analyzed using sequencing and
array-based molecular profiling approaches.</p>
<p>We extended our analysis using TumorMap to perform integrated co-clustering
analysis of the combined gene expression (n = 1,196) and DNA methylation (n =
867) profiles. Clusters in the map indicate groups of samples with high
similarity of integrated gene expression and DNA methylation profiles.</p>
<p>Ceccarelli,M., Barthel,F.P., Malta,T.M., Sabedot,T.S., Salama,S.R., Murray,B.A.,
Morozova,O., Newton,Y., Radenbaugh,A., Pagnotta,S.M., et al. (2016)
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/26824661">Molecular Profiling Reveals Biologically Discrete Subsets and Pathways of
Progression in Diffuse Glioma</a>.
Cell, 164, 550–63.</p>
</div>
<div class="section" id="quakebrain">
<span id="id2"></span><h2>QuakeBrain<a class="headerlink" href="#quakebrain" title="Permalink to this headline">¶</a></h2>
<p>We used single cell RNA sequencing on 466 cells to capture the cellular
complexity of the adult and fetal human brain at a whole transcriptome level.
Healthy adult temporal lobe tissue was obtained during surgical procedures where
otherwise normal tissue was removed to gain access to deeper hippocampal
pathology in patients with medical refractory seizures. We were able to classify
individual cells into all of the major neuronal, glial, and vascular cell types
in the brain. We were able to divide neurons into individual communities and
show that these communities preserve the categorization of interneuron subtypes
that is typically observed with the use of classic interneuron markers.</p>
<p>Darmanis,S., Sloan,S.A., Zhang,Y., Enge,M., Caneda,C., Shuer,L.M., Hayden
Gephart,M.G., Barres,B.A. and Quake,S.R. (2015)
<a class="reference external" href="http://www.pnas.org/content/112/23/7285.abstract">A survey of human brain transcriptome diversity at the single cell level</a>.
Proc. Natl. Acad. Sci. U. S. A., 112, 7285–90.</p>
</div>
<div class="section" id="pchips">
<span id="id3"></span><h2>pCHIPS<a class="headerlink" href="#pchips" title="Permalink to this headline">¶</a></h2>
<p>The presented pChips data set is a subset of Pancan12 data supplemented by
clinical tissue from lethal metastatic castration-resistant prostate cancer
patients obtained at rapid autopsy.</p>
<p>Drake,J.M., Paull,E.O., Graham,N.A., Lee,J.K., Smith,B.A., Titz,B.,
Stoyanova,T., Faltermeier,C.M., Uzunangelov,V., Carlin,D.E., et al. (2016)
<a class="reference external" href="http://www.sciencedirect.com/science/article/pii/S0092867416309138">Phosphoproteome Integration Reveals Patient-Specific Networks in Prostate
Cancer</a>.
Cell, 166, 1041–1054.</p>
</div>
<div class="section" id="mgmarin-public-pcawg-juncbase">
<span id="pcawgjuncbase"></span><h2>mgmarin_public/PCAWG_JuncBASE<a class="headerlink" href="#mgmarin-public-pcawg-juncbase" title="Permalink to this headline">¶</a></h2>
<p>TBD</p>
</div>
<div class="section" id="attribute-descriptions">
<h2>Attribute Descriptions<a class="headerlink" href="#attribute-descriptions" title="Permalink to this headline">¶</a></h2>
<div class="section" id="pancan12-samplemap-attributes">
<span id="pancan12-samplemapattributes"></span><h3>Pancan12/SampleMap Attributes<a class="headerlink" href="#pancan12-samplemap-attributes" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">Tissue</div>
<div class="line">BRCA Subtype</div>
<div class="line">COADREAD Subtype</div>
<div class="line">GBM Subtype</div>
<div class="line">OV subtype</div>
<div class="line">UCEC Subtype</div>
<div class="line">gender</div>
<div class="line">number_of_lymphnodes_positive</div>
<div class="line">colon_polyps_present</div>
<div class="line">microsatellite_instability</div>
<div class="line">metastasis_pathological_spread</div>
<div class="line">height</div>
<div class="line">weight</div>
<div class="line">age_at_initial_pathologic_diagnosis</div>
<div class="line">icd_10</div>
<div class="line">lymphovascular_invasion_present</div>
<div class="line">karnofsky_performance_score</div>
<div class="line">neoplasm_histologic_grade</div>
<div class="line">icd_o_3_site</div>
<div class="line">primary_tumor_t_stage</div>
<div class="line">lymphnode_pathologic_spread</div>
<div class="line">acute_myeloid_leukemia_calgb_cytogenetics_risk_category</div>
<div class="line">tumor_stage_and_substage</div>
<div class="line">neoplasm_disease_lymph_node_stage</div>
<div class="line">primary_tumor_pathologic_spread</div>
<div class="line">history_of_colon_polyps</div>
<div class="line">tumor_stage</div>
<div class="line">pancan subtype integrated</div>
<div class="line">pancan subtype methylation</div>
<div class="line">pancan subtype RPPA</div>
<div class="line">pancan subtype mRNA</div>
<div class="line">pancan subtype miRNA</div>
<div class="line">pancan subtype mutations</div>
<div class="line">Met vs Primary</div>
<div class="line">..._MUTATION (313 mutation flags for high-confidence mutations, where * is a gene symbol in HUGO space)</div>
<div class="line">..._AMPLIFICATION (999 gene-level or chromosomal amplification events)</div>
<div class="line">..._DELETION (1987 gene-level or chromosomal deletion events)</div>
<div class="line">TF_IPL_* (774 transcription factors with their activities summarized in the PARADIGM IPL space per each sample; * is a gene symbol in HUGO space)</div>
<div class="line">* program (42 drug programs inferred from the gene expression data, where * is a molecular process or function name)</div>
<div class="line">Mutation Signature 1</div>
<div class="line">Mutation Signature 2</div>
<div class="line">Mutation Signature 3</div>
<div class="line">Mutation Signature 4</div>
<div class="line">Mutation Signature 5</div>
<div class="line">Mutation Signature 6</div>
<div class="line">Mutation Signature 7</div>
<div class="line">Mutation Signature 8</div>
<div class="line">Mutation Signature 10</div>
<div class="line">Mutation Signature 13</div>
<div class="line">Mutation Signature 14</div>
<div class="line">Mutation Signature 17</div>
<div class="line">Mutation Signature 20</div>
<div class="line">Mutation Signature 26</div>
<div class="line">Mutation Signature 27</div>
</div>
</div>
<div class="section" id="pancan12-genemap-attributes">
<span id="pancan12-genemapattributes"></span><h3>Pancan12/GeneMap Attributes<a class="headerlink" href="#pancan12-genemap-attributes" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="26%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Database name</th>
<th class="head">Number of sets</th>
<th class="head">Variable type</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>MSigDB Positional Gene Sets</td>
<td>326</td>
<td>Binary</td>
</tr>
<tr class="row-odd"><td>MSigDB Hallmark gene sets</td>
<td>50</td>
<td>Binary</td>
</tr>
<tr class="row-even"><td>MSigDB Canonical gene sets</td>
<td>1330</td>
<td>Binary</td>
</tr>
<tr class="row-odd"><td><a class="reference external" href="GO:Biological">GO:Biological</a> Process</td>
<td>825</td>
<td>Binary</td>
</tr>
<tr class="row-even"><td><a class="reference external" href="GO:Cellular">GO:Cellular</a> Component</td>
<td>233</td>
<td>Binary</td>
</tr>
<tr class="row-odd"><td><a class="reference external" href="GO:Molecular">GO:Molecular</a> Function</td>
<td>396</td>
<td>Binary</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="getStarted.html">Get Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="basicControls.html">Basic Controls</a></li>
<li class="toctree-l1"><a class="reference internal" href="menus.html">Menus</a></li>
<li class="toctree-l1"><a class="reference internal" href="shortlist.html">Short List</a></li>
<li class="toctree-l1"><a class="reference internal" href="advancedControls.html">Advanced Controls</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Public Maps</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#pancan12">Pancan12</a></li>
<li class="toctree-l2"><a class="reference internal" href="#gliomas">Gliomas</a></li>
<li class="toctree-l2"><a class="reference internal" href="#quakebrain">QuakeBrain</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pchips">pCHIPS</a></li>
<li class="toctree-l2"><a class="reference internal" href="#mgmarin-public-pcawg-juncbase">mgmarin_public/PCAWG_JuncBASE</a></li>
<li class="toctree-l2"><a class="reference internal" href="#attribute-descriptions">Attribute Descriptions</a></li>
</ul>
</li>
</ul>

          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
            </form>
          </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="attrInfo.html" title="Attribute Information"
              >previous</a> |
            <a href="genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
              <br/>
              <a href="_sources/existingMaps.rst.txt"
                rel="nofollow">Show Source</a>
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
        &#169; Copyright 2016, UCSC Tumor Map development team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.3.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>